pyasic/miners/backends/epic.py

```diff
@@ -450,3 +455,70 @@ async def _get_pools(self, web_summary: dict = None) -> List[PoolMetrics]:
                 return pool_data
         except LookupError:
             pass
+
+    async def upgrade_firmware(self, file: Path, keepsettings: bool, password: str):
+
+        """
+        Upgrade the firmware of the ePIC miner device.
+
+        Args:
+            file (Path): The local file path of the firmware to be uploaded.
+            keepsettings (bool): Whether to keep the current settings after the update.
+            password (str): The password for authentication.
+
+        Returns:
+            str: Confirmation message after upgrading the firmware.
+        """
+        try:
+            logging.info("Starting firmware upgrade process for ePIC miner.")
+
+            if not file:
+                raise ValueError("File location must be provided for firmware upgrade.")
+
+            # calculate the SHA256 checksum of the firmware file
+            sha256_hash = hashlib.sha256()
+            async with aiofiles.open(file, "rb") as f:
+                while chunk := await f.read(8192):
+                    sha256_hash.update(chunk)
+            checksum = sha256_hash.hexdigest()
+
+            # prepare the multipart/form-data request
+            form_data = aiohttp.FormData()
+            form_data.add_field('checksum', checksum)
+            form_data.add_field('keepsettings', str(keepsettings).lower())
+            form_data.add_field('password', password)
+            form_data.add_field('update.zip', open(file, 'rb'), filename='update.zip')
+
+            # Send the POST request to the ePIC miner device
+            async with aiohttp.ClientSession() as session:
```

Comment:
Use the web attribute instead of doing this explicitly.

Reply:
I have updated the code to use the `web` attribute.